File: leetcode/integer-to-roman.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    string intToRoman(int num) {
        string result = "";

        while (num > 0) {
            // 坑1: 相等的情况
            if (num >= 1000) {
                num -= 1000;
                result += "M";
            } else if (num >= 900) {
                num -= 900;
                // 坑2: 位置写反了
                result += "CM";
            } else if (num >= 500) {
                num -= 500;
                result += "D";
            } else if (num >= 400) {
                num -= 400;
                result += "CD";
            } else if (num >= 100) {
                num -= 100;
                result += "C";
            } else if (num >= 90) {
                num -= 90;
                result += "XC";
            } else if (num >= 50) {
                num -= 50;
                result += "L";
            } else if (num >= 40) {
                num -= 40;
                result += "XL";
            } else if (num >= 10) {
                num -= 10;
                result += "X";
            } else if (num >= 9) {
                num -= 9;
                result += "IX";
            } else if (num >= 5) {
                num -= 5;
                result += "V";
            } else if (num >= 4) {
                num -= 4;
                result += "IV";
            } else {
                num -= 1;
                result += "I";
            }
        }

        return result;
    }
};

int main() {

    Solution s = Solution();
    cout << s.intToRoman(3) << endl;
    cout << s.intToRoman(4) << endl;
    cout << s.intToRoman(500) << endl;
    cout << s.intToRoman(3999) << endl;
    cout << s.intToRoman(999) << endl;
    cout << s.intToRoman(199) << endl;

    return 0;
}

File: leetcode/IntegerToRoman.java
package leetcode;

/**
 * IntegerToRoman
 * https://leetcode-cn.com/problems/integer-to-roman/
 * 12. 整数转罗马数字
 * https://leetcode-cn.com/problems/integer-to-roman/solution/cong-da-dao-xiao-guo-lu-bing-cun-chu-jie-0yeh/
 * @author tobin
 * @since 2021-05-14
 */
public class IntegerToRoman {
    public static void main(String[] args) {
        IntegerToRoman sol = new IntegerToRoman();

        System.out.println(sol.intToRoman(3));
        System.out.println(sol.intToRoman(4));
        System.out.println(sol.intToRoman(9));
        System.out.println(sol.intToRoman(58));
        System.out.println(sol.intToRoman(1994));
    }

    private static final int[] ranges = {
            1000, 900, 500, 400,
            100, 90, 50, 40,
            10, 9, 5, 4,
            1};
    private static final String[] symbols = {
            "M", "CM", "D", "CD",
            "C", "XC", "L", "XL",
            "X", "IX", "V", "IV",
            "I"};

    public String intToRoman(int num) {
        int value = num;
        int baseIdx = 0;
        StringBuilder resultBuilder = new StringBuilder();
        while (value > 0) {
            if (value >= ranges[baseIdx]) {
                value -= ranges[baseIdx];
                resultBuilder.append(symbols[baseIdx]);
            } else {
                baseIdx++;
            }
        }
        return resultBuilder.toString();
    }
}
